art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py

```diff
@@ -0,0 +1,447 @@
+        :type eps: `float`
+        :param eps_step: Attack step size (input variation) at each iteration.
+        :type eps_step: `float`
+        :param random_init: Random initialisation within the epsilon ball. For random_init=False
+            starting at the original input.
+        :type random_init: `bool`
+        :return: Adversarial examples.
+        :rtype: `Tensor`
+        """
+        if random_init:
+            n = x.shape[0]
+            m = np.prod(x.shape[1:])
+            x_adv = x.astype(ART_NUMPY_DTYPE) + (
+                random_sphere(n, m, eps, self.norm).reshape(x.shape).astype(ART_NUMPY_DTYPE)
+            )
+
+            if hasattr(self.estimator, "clip_values") and self.estimator.clip_values is not None:
+                clip_min, clip_max = self.estimator.clip_values
+                x_adv = np.clip(x_adv, clip_min, clip_max)
+            x_adv = tf.convert_to_tensor(x_adv)
+
+        else:
+            if isinstance(x, np.ndarray):
+                x_adv = tf.convert_to_tensor(x.astype(ART_NUMPY_DTYPE))
+            else:
+                x_adv = x
+
+        # Get perturbation
+        perturbation = self._compute_perturbation(x_adv, y, mask)
+
+        # Apply perturbation and clip
+        x_adv = self._apply_perturbation(x_adv, perturbation, eps_step)
+
+        # Do projection
+        perturbation = self._projection(x_adv - x_init, eps, self.norm)
+
+        # Recompute x_adv
+        x_adv = tf.add(perturbation, x_init)
+
+        return x_adv
+
+    @staticmethod
+    def _projection(values, eps, norm_p):
+        """
+        Project `values` on the L_p norm ball of size `eps`.
+
+        :param values: Values to clip.
+        :type values: `Tensor`
+        :param eps: Maximum norm allowed.
+        :type eps: `float`
+        :param norm_p: L_p norm to use for clipping. Only 1, 2 and `np.Inf` supported for now.
+        :type norm_p: `int`
+        :return: Values of `values` after projection.
+        :rtype: `Tensor`
+        """
+        # Pick a small scalar to avoid division by 0
+        tol = 10e-8
+        values_tmp = tf.reshape(values, (values.shape[0], -1))
+
+        if norm_p == 2:
+            values_tmp = values_tmp * tf.expand_dims(
+                tf.minimum(1.0, eps / (tf.norm(values_tmp, ord=2, axis=1) + tol)), axis=1
+            )
+
+        elif norm_p == 1:
+            values_tmp = values_tmp * tf.expand_dims(
+                tf.minimum(1.0, eps / (tf.norm(values_tmp, ord=1, axis=1) + tol)), axis=1
+            )
+
+        elif norm_p == np.inf:
+            values_tmp = tf.sign(values_tmp) * tf.minimum(tf.math.abs(values_tmp), eps)
+
+        else:
+            raise NotImplementedError(
+                "Values of `norm_p` different from 1, 2 and `np.inf` are currently not supported."
+            )
+
+        values = tf.reshape(values_tmp, values.shape)
+
+        return values
+
+    def set_params(self, **kwargs):
```

Comment:
See comment about set_params for `projected_gradient_descent_pytorch.py`.

Reply:
No, there are no inheritance relationship among the attacks.

Reply:
Yes, this is done with a new parent class for all implementation of PGD.